src/core/lib/event_engine/posix_engine/timer_manager.cc

```diff
@@ -155,54 +72,32 @@ void TimerManager::MainLoop() {
     grpc_core::Timestamp next = grpc_core::Timestamp::InfFuture();
     absl::optional<std::vector<experimental::EventEngine::Closure*>>
         check_result = timer_list_->TimerCheck(&next);
-    if (check_result.has_value()) {
-      if (!check_result->empty()) {
-        RunSomeTimers(std::move(*check_result));
-        continue;
-      }
-    } else {
-      /* This case only happens under contention, meaning more than one timer
-         manager thread checked timers concurrently.
-
-         If that happens, we're guaranteed that some other thread has just
-         checked timers, and this will avalanche into some other thread seeing
-         empty timers and doing a timed sleep.
-
-         Consequently, we can just sleep forever here and be happy at some
-         saved wakeup cycles. */
-      next = grpc_core::Timestamp::InfFuture();
+    GPR_ASSERT(check_result.has_value() &&
+               "ERROR: More than one MainLoop is running.");
+    if (!check_result->empty()) {
+      RunSomeTimers(std::move(*check_result));
+      continue;
     }
-    if (!WaitUntil(next)) return;
-  }
-}
-
-void TimerManager::RunThread(void* arg) {
-  g_timer_thread = true;
-  std::unique_ptr<RunThreadArgs> thread(static_cast<RunThreadArgs*>(arg));
-  if (grpc_event_engine_timer_trace.enabled()) {
-    gpr_log(GPR_DEBUG, "TimerManager::%p starting thread::%p", thread->self,
-            &thread->thread);
-  }
-  thread->self->Run();
-  if (grpc_event_engine_timer_trace.enabled()) {
-    gpr_log(GPR_DEBUG, "TimerManager::%p thread::%p finished", thread->self,
-            &thread->thread);
+    if (!WaitUntil(next)) break;
   }
-}
-
-void TimerManager::Run() {
-  MainLoop();
-  grpc_core::MutexLock lock(&mu_);
-  thread_count_--;
-  if (thread_count_ == 0) cv_threadcount_.Signal();
+  main_loop_exit_signal_.Notify();
 }
 
 bool TimerManager::IsTimerManagerThread() { return g_timer_thread; }
 
-TimerManager::TimerManager() : host_(this) {
+TimerManager::TimerManager(
+    std::shared_ptr<grpc_event_engine::experimental::ThreadPool> thread_pool)
+    : host_(this), thread_pool_(std::move(thread_pool)) {
   timer_list_ = std::make_unique<TimerList>(&host_);
-  grpc_core::MutexLock lock(&mu_);
-  StartThread();
+  main_thread_ = grpc_core::Thread(
```

Comment:
I think this should be `thread_pool->Run([this] { MainLoop(); });`

Reply:
It is intentionally not on the thread pool. There is now just one persistent TimerManager main loop. In the current incarnation, we do not want it executing closures, timer or otherwise, when work stealing is landed (presuming that happens before the timer manager's polling is integrated with the I/O polling thread, which I will do as soon as Vignesh's work on the poller https://github.com/grpc/grpc/pull/31128 is merged and stable).

Alternative solutions could exist here that let a TimerManager with multiple MainLoops run on the thread pool, with a fair amount of added complexity. This PR is an intermediate step that allows work stealing to influence timer callbacks. I don't necessarily think it's worth managing a pool within a pool to allow callbacks to execute in the same thread, especially when that code is going away very quickly.

Reply:
I wasn't suggesting a pool within a pool, just that we have one MainLoop running inside the thread pool, instead of firing up a thread. My intent there is to ensure we've resolved any memory management hazards here (of the form 'who owns who') before moving onto the next step.

No need to block on this though, useful discussion.

Reply:
> No need to block on this though, useful discussion.

Agreed! Thanks. Updating this with our chat: the fact that the MainLoop is in a non-thread-pool thread makes it harder to have those shared_ptr cycles and shutdown issues. We'll have to consider them again when the MainLoop thread merges into the poller thread(s) on the thread pool.